### PR TITLE
Implement repo fetcher (historical data)

### PR DESCRIPTION
**Pull request by dennwc**
Implement repository metadata fetcher for historical data. This cloud function:
1) Intercepts repository update events sent by the App installation handler (see #13).
2) Pulls the latest metadata from the Github API.
3) Updates repository metadata in the DB.
4) Sends this metadata state to a repository topic (for the archival or other processing).
5) Triggers updates for issues, PRs, releases, branches, etc. No handlers for those are defined yet.

Note that this function won't use the current DB schema. Instead a different schema is defined with different assumptions in mind:
- Only the latest state is stored. Event-driven data we care about is only related to PRs.
- IDs are implemented as proper foreign keys.
- Aggregate fields are removed. SQL can aggregate things on its own.
- Things are never deleted. We will suppress events for now, and set a `deleted_at` flag later.

It also adds a second helper cloud function which has no inputs wire to other function, but can be triggered manually to initiate a cascading re-fetch of all repos, PRs, etc in a specific installation.

### CODE REVIEW

File: github/database.go [hunk 1]
```diff
@@ -588,3 +624,172 @@ func (db *Database) DeleteInstallation(ctx context.Context, installID int64) err
...
+func boolField(f *bool) interface{} {
+	if f == nil {
+		return nil
+	}
+	return *f
+}
+
+func timeField(f *gh.Timestamp) interface{} {
+	if f == nil {
+		return nil
```
Comment (vmarkovtsev) [inline github/database.go]: Good old Go with 100500 repetitive functions for different data types...

File: github/database.go [hunk 2]
```diff
@@ -588,3 +624,172 @@ func (db *Database) DeleteInstallation(ctx context.Context, installID int64) err
...
+	}
+	return f.UTC()
+}
+
+var repoFields = []struct {
+	Col string
+	Fnc func(r *gh.Repository) interface{}
```
Comment (vmarkovtsev) [inline github/database.go]: "Fnc" looks Rust-y... How about naming them `function` and `Function`?

File: github_repos.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+	"fmt"
+	"log"
+
+	"github.com/athenianco/metadata/github"
+	"github.com/athenianco/metadata/pubsub"
+	gh "github.com/google/go-github/v28/github"
+)
+
+// GithubRepoUpdater triggers on repo update events.
+// It fetches the latest repo state from the Github, updates it in the database and sends a copy to GITHUB_REPOS_TOPIC.
```
Comment (vmarkovtsev) [inline github_repos.go]: OK, we should also document that we need to define `GITHUB_REPOS_TOPIC`

File: github_repos.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+	}
+	h.prs, err = pubsub.NewPublisherFromEnv("GITHUB_REPO_PRS_TOPIC")
+	if err != nil {
+		return err
+	}
+	h.refs, err = pubsub.NewPublisherFromEnv("GITHUB_REPO_REFS_TOPIC")
+	if err != nil {
+		return err
+	}
+	h.releases, err = pubsub.NewPublisherFromEnv("GITHUB_REPO_RELEASES_TOPIC")
```
Comment (vmarkovtsev) [inline github_repos.go]: ... and the rest of these envvars.

File: github_repos.go [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+func (h *GithubRepoHelper) handleReloadInstallationRepos(ctx context.Context, installID int64) error {
+	log.Printf("Reloading metadata for repos on installation %d", installID)
+	repos, err := h.db.GetInstallationRepos(ctx, installID)
+	if err != nil {
+		return err
+	} else if len(repos) == 0 {
+		return errors.New("no repos enabled on this installation")
+	}
+	log.Printf("Sending metadata update on %d repos for installation %d", len(repos), installID)
```
Comment (vmarkovtsev) [inline github_repos.go]: BTW Stackdriver automatically parses JSON in stdout/stderr, so we should consider switching to structured logging in metadata.

File: schema.sql [hunk 6]
```diff
@@ -53,6 +53,56 @@ CREATE TABLE public.github_installation_repos (
...
 );
 
+--
+-- Name: github_users_v2; Type: TABLE; Schema: public;
+--
+
+CREATE TABLE public.github_users_v2 (
+    id bigint PRIMARY KEY,
+    node_id text UNIQUE,
+    login text NOT NULL, -- it should be unique, but we relax this requirement because event processing forces eventual consistency
```
Comment (vmarkovtsev) [inline schema.sql]: What if a user changes the name? From https://help.github.com/en/github/setting-up-and-managing-your-github-user-account/changing-your-github-username

> After changing your username, your old username becomes available for anyone else to claim.

File: schema.sql [hunk 7]
```diff
@@ -350,7 +400,7 @@ CREATE TABLE public.github_users (
     following bigint,
     hireable boolean,
     htmlurl text,
-    id bigint NOT NULL,
+    id bigint PRIMARY KEY,
```
Comment (vmarkovtsev) [inline schema.sql]: This matches https://caius.github.io/github_id/ doesn't it? (ensuring to avoid integration problems)

File: Makefile [hunk 8]
```diff
@@ -31,17 +31,29 @@ GITHUB_PROCESSOR_ENTRY_POINT = "GithubProcessFunc"
...
 
 GITHUB_ARCHIVER_NAME ?= "github_archive-owl"
-GITHUB_ARCHIVER_ENTRY_POINT = "GithubArchiveFunc"
+GITHUB_ARCHIVER_ENTRY_POINT = "GithubEventsArchiveFunc"
 GITHUB_ARCHIVER_BUCKET = "athenian-owl-github-events"
 
 GITHUB_REPOS_UPDATE_TOPIC = "github-repo-update"
 GITHUB_ORGS_UPDATE_TOPIC = "github-org-update"
+GITHUB_REPOS_TOPIC = "github-repo-meta"
+GITHUB_REPOS_HELPER_TOPIC = "github-repo-cmd"
```
Comment (se7entyse7en) [inline Makefile]: This could be interesting to expose in a backoffice section in the future for manual re-trigger.

File: github_repos.go [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+		return err
+	}
+	return nil
+}
+
+// triggerListUpdates triggers repo items list updates for issues, PRs, refs, releases.
+func (h *GithubRepoUpdater) triggerListUpdates(ctx context.Context, r *gh.Repository) error {
+	ev := ghRepoEvent{
+		Repository: r,
+		EventType:  ghRepoInstall,
```
Comment (se7entyse7en) [inline github_repos.go]: Shouldn't this depend on the [actual event type](https://github.com/athenianco/metadata/pull/15/commits/caededaddbbd087f4fd6079ecb30bb8b80c0fee2?file-filters%5B%5D=.go#diff-4fbabc8295eae08b186b4a0ae8fea20aR71) that triggered these updates?

File: github_repos.go [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+		repos, err := h.db.GetInstallationRepos(ctx, m.InstallID)
+		if err != nil {
+			return err
+		} else if len(repos) == 0 {
+			return errors.New("no repos enabled on this installation")
+		}
+		log.Printf("Sending metadata update on %d repos for installation %d", len(repos), m.InstallID)
+		events := make([]interface{}, 0, len(repos))
+		for _, id := range repos {
+			id := id
```
Comment (se7entyse7en) [inline github_repos.go]: Could you explain me this?

**Comment by se7entyse7en** [comment]
Also AFAIS with this PR we're already switching user and repos tables. I guess that before the launch we just nuke everything and start the processing from scratch.

**Comment by se7entyse7en** [comment]
> Note that this function won't use the current DB schema. Instead a different schema is defined with different assumptions in mind:

> - Only the latest state is stored. Event-driven data we care about is only related to PRs.

This means that we, for example, keep only the latest value of `updated_at` right?

> - IDs are implemented as proper foreign keys.
> - Aggregate fields are removed. SQL can aggregate things on its own.
> - Things are never deleted. We will suppress events for now, and set a deleted_at flag later.

Could you elaborate this please?

File: github/client.go [hunk 11]
```diff
@@ -26,6 +26,13 @@ func NewClient(clientID, clientSecret string) *Client {
 // NewClientFromEnv is similar to NewClient, but get's required parameters from GITHUB_CLIENT_ID and GITHUB_CLIENT_SECRET
 // environment variables.
 func NewClientFromEnv() (*Client, error) {
+	if token := os.Getenv("GITHUB_TOKEN"); token != "" {
```
[the remark this answers is not in the document]
Comment (dennwc) [reply github/client.go]: It's only required for some services and it's "documented" in the `Makefile` (since we don't have a good way to deploy those functions for now).

We will soon define a better way how to define cloud functions and generate those dependencies/docs for it. So I'd prefer to keep it as-is for now, if you don't mind.

File: github/database.go [hunk 12]
```diff
@@ -509,6 +516,11 @@ func (db *Database) CreateInstallation(ctx context.Context, delivery string, ins
 		inst.GetCreatedAt().UTC(), inst.GetUpdatedAt().UTC(),
 		inst.GetHTMLURL(),
 	)
+	if err != nil {
+		_ = tx.Rollback()
```
[the remark this answers is not in the document]
Comment (dennwc) [reply github/database.go]: Do we care about those though? Rollback errors would mostly happen if the connection dies, but in this case the server should rollback anyway.

File: github_repos.go (hunk 5 above)
Comment (dennwc) [reply github_repos.go]: Opened an issue for it:
https://github.com/athenianco/metadata/issues/17

File: schema.sql (hunk 6 above)
Comment (dennwc) [reply schema.sql]: Current schema allows renames. Am I missing something?

File: schema.sql (hunk 7 above)
Comment (dennwc) [reply schema.sql]: Correct, those are native Github user IDs. This is why I defined it as `bigint` instead of `autoincrement`.

File: Makefile (hunk 8 above)
Comment (dennwc) [reply Makefile]: This does exactly this - I was using it to re-run the pipeline during testing :)

File: github_repos.go (hunk 9 above)
Comment (dennwc) [reply github_repos.go]: It might, but I wanted to narrow down the set of types that those functions switch on. Those are still TBD, so for now most of those set the type to "install" instead of passing the original one.

File: github/client.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply github/client.go]: Just mention it in the README please, because this is what the other two parties do.

File: github/database.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply github/database.go]: This is about logging the problems, and I care about logging, for example.

File: github_repos.go (hunk 10 above)
Comment (dennwc) [reply github_repos.go]: This code intentionally shadows an `id` variable from the loop with a new local variable with the same name.

It works similar to this:
```go
{
 var id int
 for id = range repos {
  var id int = id
  // ...
 }
}
```
(loop refines it's own scope in `range` that is different from the inner scope of the loop)

This code of course can be rewritten as:

```go
for id := range repos {
  id2 := id
  // use id2 instead of id
}
```

The reason why I shadowed the variable is this line:

```go
  Repository: &gh.Repository{ID: &id},
```

We must declare a new variable because in other case the line above will capture the address of the loop variable and all those `gh.Repository` objects will have the same ID (the last one in the array, because this is the last value assigned by the loop).

Similar bug usually happens when devs write a loop which spawns goroutines:

```go
for id := range repos {
  go func(){
    // use id here
  }()
}
```

This is very dangerous because it's hard to notice and it acts exactly as I described above: all routines will get the same ID.

So to prevent those kind of bugs I usually define a new variable and shadow the old one to prevent other devs from accidentally using the wrong one in the future (`id := id` line).

File: github/database.go [hunk 13]
```diff
@@ -588,3 +624,172 @@ func (db *Database) DeleteInstallation(ctx context.Context, installID int64) err
 	}
 	return tx.Commit()
 }
+
+// upsertRepos ensures that a bare minimum information for repositories is available in the database.
+// It acts an an upsert - if there is no such repo, it will create it with a few basic fields, and if it's already there
+// it will do nothing.
+func (db *Database) upsertRepos(ctx context.Context, tx *sql.Tx, repos []*gh.Repository) error {
+	stmt, err := tx.PrepareContext(ctx,
+		`INSERT INTO github_repositories_v2(id, node_id, name, private) VALUES ($1, $2, $3, $4)
```
[the remark this answers is not in the document]
Comment (dennwc) [reply github/database.go]: Correct. I don't want to change old ones because it will cause all kinds of issues with migrations and will probably block your work on the API. Maybe in the transition we will have to query both tables, but it should be easy to match because of those GH IDs.

File: github/client.go [hunk 14]
```diff
@@ -41,17 +48,20 @@ func NewClientFromEnv() (*Client, error) {
...
 		commits []*gh.RepositoryCommit
 		resp    *gh.Response
 	)
-	pr := event.GetPullRequest()
-	repo := event.GetRepo()
-	owner := repo.GetOwner().GetLogin()
 	opt := &gh.ListOptions{
 		PerPage: 20,
 	}
+	owner := repo.GetOwner().GetLogin()
```
[the remark this answers is not in the document]
Comment (dennwc) [reply github/client.go]: We can solve it in two ways:

1) Change the existing schema, migrate, redeploy
2) Wait for the same kind of PR that defines `pull_request_commits_v2`

Also related to https://github.com/athenianco/metadata/pull/15#discussion_r375528289

File: schema.sql (hunk 6 above)
Comment (vmarkovtsev) [reply schema.sql]: :100: How about creating a JIRA issue to remember about it? Or at least inserting a TODO...

File: github_repos.go (hunk 4 above)
Comment (dennwc) [reply github_repos.go]: I'm working on a way to generate docs and deploy scripts from something like this:

```yml
functions:
  - name: github-hook
    handler: GithubWebhook
    trigger:
      type: http
    vars:
      - GITHUB_WEBHOOK_SECRET_KEY

  - name: github_archive
    handler: GithubEventsArchive
    trigger:
      type: pubsub
      topic: github-hook
    vars:
      - GITHUB_EVENTS_BUCKET
```

It will include all those custom envs, and some of them will be assigned automatically (e.g. `github-hook` topic becomes `GITHUB_HOOK_TOPIC` env).

File: github_repos.go (hunk 10 above)
Comment (vmarkovtsev) [reply github_repos.go]: (side note: we realized in the ML team that whenever somebody asks for clarifications during the review, that spot is a good candidate for inserting comments in the code)

File: github/client.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply github/client.go]: I would say that both options work for me.

File: github_repos.go (hunk 4 above)
Comment (vmarkovtsev) [reply github_repos.go]: That would be very nice and helpful :+1:

**Comment by dennwc** [comment]
@se7entyse7en 

> Also AFAIS with this PR we're already switching user and repos tables.

Not yet. This is why I defined tables with a `_v2` suffix: we can avoid migration for now and see if we are missing anything in the new schema. But in general you are right, it's better to trigger those functions and let them do the work instead of migrating data with scripts.

> This means that we, for example, keep only the latest value of `updated_at` right?

Correct. However, I also plan to store a separate table with events that refers to the same repo ID as a FK. But we'll see if it's really required or not.

> Could you elaborate this please?

Could you elaborate on the question? :sweat_smile: Do you mean FKs, aggregates, deletes, or what exactly?

In short, old schema doesn't have FK, so no dataset validation is possible. New schema fixes it.

By aggregates I mean counters like `stars`, `forks`, etc. If we pull all the data regarding those, we can `COUNT` them instead. If we need a cached value, we can define a materialized view instead.

Regarding deletes, I'd prefer to keep repo records around since IDs won't ever repeat. This will eventually allow to reference to deleted PRs/issues/repo from other places, while still keeping those values a real FKs. Instead of deleting we can set a `deleted_at` field to mark the record as "dead".

File: github/database.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (dennwc) [reply github/database.go]: But again, why do you care about the timeout on the transaction cancellation, if the main cause was different?

[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply github/database.go]: It doesn't matter, all the errors should be logged.

[the remark this answers is not in the document]
Comment (dennwc) [reply github/database.go]: This is just a statement without any reasoning behind it. Can you elaborate why?

File: github/client.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (dennwc) [reply github/client.go]: Done

File: github_repos.go (hunk 10 above)
Comment (dennwc) [reply github_repos.go]: Added a comment

File: github/database.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply github/database.go]: You mean I shall explain why good logs are good?

[the remark this answers is not in the document]
Comment (dennwc) [reply github/database.go]: "Good" is subjective. IMO, printing the second error only increases the noise and won't add any useful information. You thinking is different, so please give a practical example where the second error gives any additional information comparing to the first one.

File: github/client.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (se7entyse7en) [reply github/client.go]: I'd say that maybe option 2 is better, but up to you.

File: github/database.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (se7entyse7en) [reply github/database.go]: I agree with both of you in the sense that having noise for sure doesn't help to find useful information,  but at the same time I think that in these cases it's important to have the full context of what happened. Moreover, we're talking about adding more information to the same log line and not about adding extra logs, so I don't think it's going to be that much noisy. Actually, for me it would be even fine if they're logged in different lines.

Regarding a practical example, tell me if I'm wrong, but I think that knowing that a rollback has failed would at least make me check what was part of the transaction. This means that potentially the database is in a dirty state depending on the reason why the rollback failed.

@dennwc as you said (emphasis is mine):

> Rollback errors would mostly happen if the connection dies, but in this case the server _**should**_ rollback anyway.

But I don't know all the possible scenarios a rollback could fail (concurrent writes? a migration during the transaction? a failover process in-between? etc.) So I think that it would be valuable to have that information as well considering that it's not gonna make the logs much dirtier.

File: schema.sql (hunk 6 above)
Comment (se7entyse7en) [reply schema.sql]: @vmarkovtsev you mean that we should add the logic for reacting to renaming?

File: Makefile (hunk 8 above)
Comment (se7entyse7en) [reply Makefile]: Yup, I mean that it would be great in near-future to have a section in the webapp with a button that triggers it for super-admins (us) 👍

File: github_repos.go (hunk 10 above)
Comment (se7entyse7en) [reply github_repos.go]: Ah ok, got it! Thanks!

**Comment by se7entyse7en** [comment]
> it's better to trigger those functions and let them do the work instead of migrating data with scripts.

In general, I agree, whenever we can avoid data migrations the better, especially with running ETL pipeline. But how to properly handle schema changes in the future is a bigger topic that we can discuss in a second step. 👍 

> But we'll see if it's really required or not.

➕yes please, let's see first if required 👍 

> Regarding deletes, I'd prefer to keep repo records around since IDs won't ever repeat. This will eventually allow to reference to deleted PRs/issues/repo from other places, while still keeping those values a real FKs. Instead of deleting we can set a deleted_at field to mark the record as "dead".

I didn't understand what you mean with:

>  We will suppress events for now

But I completely agree with the policy of not deleting anything. Adding a `deleted_at` field is the best option for me as well 💯
